mail.go

```diff
@@ -0,0 +1,284 @@
+		m.addHeader("To", addr)
+	}
+	return m
+}
+
+func (m *Message) SetFrom() *Message {
+	if m.err != nil {
+		return m
+	}
+
+	m.setHeader("From", m.senderName+" <"+m.fromAddress+">")
+	return m
+}
+
+func (m *Message) SetSubject() *Message {
+	if m.err != nil {
+		return m
+	}
+
+	m.setHeader("Subject", m.subject)
+	return m
+}
+
+func (m *Message) AddTo(addresses ...string) *Message {
+	if m.err != nil {
+		return m
+	}
+
+	m.addRecipientAddresses(addresses...)
+
+	return m
+}
+
+// create a multipart MIME message with text and file attachment
+func (m *Message) SetBody(multipartType string, f *excelize.File) *Message {
+	if m.err != nil {
+		return m
+	}
+
+	var header textproto.MIMEHeader
+
+	// create a new multipart writer
+	writer := multipart.NewWriter(m.body)
+	contentType := "multipart/" + multipartType + ";\n \tboundary=" + writer.Boundary()
+	// add header to main header group
+	m.setHeader("Content-Type", contentType)
+
+	// add text part
+	textBody := "\r\nPlease see attached file with portal accounts.\r\n"
+	header = make(textproto.MIMEHeader)
+	header.Set("Content-Type", "text/plain; charset="+charSet)
+	header.Set("Content-Transfer-Encoding", bodyEncoding)
+	p, err := writer.CreatePart(header)
+	if err != nil {
+		m.err = err
+		return m
+	}
+	_, err = p.Write([]byte(b64.StdEncoding.EncodeToString([]byte(textBody))))
+	if err != nil {
+		m.err = err
+		return m
+	}
+
+	// add file attachment part
+	ft, err := newFile(f)
+	if err != nil {
+		m.err = err
+		return m
+	}
+
+	header = make(textproto.MIMEHeader)
+	value := fmt.Sprintf("%s;\n \tname=%q", ft.mimeType, ft.name)
+	header.Set("Content-Type", value)
+	header.Set("Content-Transfer-Encoding", bodyEncoding)
+	value = fmt.Sprintf("attachment;\n \tfilename=%q", ft.name)
+	header.Set("Content-Disposition", value)
+	p, err = writer.CreatePart(header)
+	if err != nil {
+		m.err = err
+		return m
+	}
+	_, err = p.Write([]byte(b64.StdEncoding.EncodeToString(ft.data)))
+	if err != nil {
+		m.err = err
+		return m
+	}
+
+	err = writer.Close()
+	if err != nil {
+		m.err = err
+		return m
+	}
+
+	return m
+}
+
+func (m *Message) MIMEString() string {
+	return m.getHeaders() + m.body.String()
+}
+
+func (m *Message) getHeaders() (headers string) {
+	m.setHeader("Date", time.Now().Format(time.RFC1123Z))
+
+	// combine the headers
+	for header, values := range m.headers {
+		headers += header + ": " + strings.Join(values, ", ") + "\r\n"
+	}
+
+	headers = headers + "\r\n"
+
+	return
+
+}
+
+func (m *Message) validateRecipientAddresses() *Message {
+	// remove duplicate addresses
+	// log invalid address and continue
+	// if no valid address, set m.err
+	addresses := make([]string, 0)
+	addressSet := make(map[string]bool)
+	for _, addr := range m.toAddresses {
+		if len(addr) > 0 {
+			a, err := mail.ParseAddress(addr)
+			if err != nil {
+				log.Printf("validating recipient address %s: %s", addr, err)
+				continue
+			}
+			if _, ok := addressSet[a.Address]; !ok {
```

Comment:
why do you need to deduplicate? does something fail if you don't? do you have reason to suspect that there will be duplicates?

Reply:
I was inspired by the go mail library that removed dups. In fact, dups are removed by something between the CMS mail relay and the recipient's inbox. Turns out I do not need to dedup. I have removed this code.